src/protocol/context.rs

```diff
@@ -10,13 +13,15 @@ use super::{securemul::SecureMul, sort::reveal::Reveal, RecordId, Step};
 pub struct ProtocolContext<'a, G, S: SpaceIndex> {
     pub participant: &'a Participant<S>,
     pub gateway: &'a G,
+    pub identity: Identity,
```

Comment:
That would indeed be pretty cute... but I am really concerned about the potential to introduce super difficult to debug problems. For example, if the identity switched halfway through modulus conversion (between XOR 1 and XOR 2) it would certainly break the logic, even without the optimisation.

The easiest way to fix the asymmetry that optimized-modulus-conversion will introduce, would be to randomly assign the roles (H1, H2, H3) *per-record*. 

Could we provide a different "ProtocolContext" to the "modulus conversion" than we do to the next step? If we are willing to create a new ProtocolContext object within the IPA protocol, we can keep this approach. That would also avoid the situation where the identity is switching around mid-protocol (which will certainly break all protocols, including Richa's shuffle).

Reply:
Oh, I see.

So basically, we could have shared randomness (across all 3 helpers) that they could consult to decide "For this particular modulus conversion, which role should I assume?). Is that what you mean?

Reply:
Or even better, just use (recordId % 3)

Reply:
My view was that we might try to balance out the networking load by alternating between clockwise sending and anti-clockwise sending for the multiplication.  That doesn't mean `identity = record_id %3`, but more `direction = record_id & 1`.

Reply:
I would follow up with removing `identity()` function